diag_manager/fms_diag_object.F90

```diff
@@ -648,43 +649,58 @@ subroutine fms_diag_send_complete(this, time_step)
+    if(.not. diag_field%is_registered()) cycle
+    if(DEBUG_SC) call mpp_error(NOTE, "fms_diag_send_complete:: var: "//diag_field%get_varname())
+    ! get files the field is in
+    allocate (file_ids(size(diag_field%get_file_ids() )))
+    file_ids = diag_field%get_file_ids()
+    math = diag_field%get_math_needs_to_be_done()
+    ! if doing math loop through each file for given field
+    if (size(file_ids) .ge. 1 .and. math) then
+      file_loop: do ifile = 1, size(file_ids)
+        diag_file => this%FMS_diag_files(ifile)
+        ! if the file is not allocated go away
+        if(.not. allocated(diag_file%FMS_diag_file)) then
+          if(DEBUG_SC) call mpp_error(NOTE, "file id:"//string(ifile)//" not allocated for field:"//diag_field%get_varname())
+          cycle
+        endif
         ! If the field is not registered go away
-        if (.not. diag_file%FMS_diag_file%is_field_registered(ifield)) cycle
-
-        diag_field => this%FMS_diag_fields(file_field_ids(ifield))
-        !> Check if math needs to be done
-        math = diag_field%get_math_needs_to_be_done()
-        calling_math: if (math) then
+        if (.not. diag_file%FMS_diag_file%is_field_registered(ifield)) then 
+          if(DEBUG_SC) call mpp_error(NOTE, "file id:"//string(ifile)//" not registered for field:"//diag_field%get_varname())
+          cycle
+        endif
+        ! Check if buffer alloc'd
+        has_input_buff: if (diag_field%has_input_data_buffer()) then
           input_data_buffer => diag_field%get_data_buffer()
+          ! reset bounds, allocate output buffer, and update it with reduction
           call bounds%reset_bounds_from_array_4D(input_data_buffer)
-          call this%allocate_diag_field_output_buffers(input_data_buffer, file_field_ids(ifield))
-          error_string = this%fms_diag_do_reduction(input_data_buffer, file_field_ids(ifield), &
-            diag_field%get_mask(), diag_field%get_weight(), &
-            bounds, .False., Time=this%current_model_time)
+          call this%allocate_diag_field_output_buffers(input_data_buffer, ifield)
+          error_string = this%fms_diag_do_reduction(input_data_buffer, ifield, &
+                              diag_field%get_mask(), diag_field%get_weight(), &
+                              bounds, .False., Time=this%current_model_time)
           if (trim(error_string) .ne. "") call mpp_error(FATAL, "Field:"//trim(diag_field%get_varname()//&
-            " -"//trim(error_string)))
-        endif calling_math
-        !> Clean up, clean up, everybody everywhere
-        if (associated(diag_field)) nullify(diag_field)
-      enddo field_loop
-      !> Clean up, clean up, everybody do your share
-      if (allocated(file_field_ids)) deallocate(file_field_ids)
-    endif field_outer_if
-  enddo file_loop
-
-  call this%fms_diag_do_io()
+                                                         " -"//trim(error_string)))
+        endif has_input_buff 
```

Comment:
Can we add an else here and error out with a message saying that `math_needs_to_be_done` is true, but there is no input buffer?